buffer-api/src/main/java/io/netty/buffer/api/bytebuffer/NioBuffer.java

```diff
@@ -1190,4 +1196,59 @@ private IndexOutOfBoundsException outOfBounds(int index) {
     ByteBuffer recoverable() {
         return base;
     }
+
+    // <editor-fold name="BufferIntegratable methods">
+    private ByteBufAdaptor adaptor;
+    @Override
+    public ByteBuf asByteBuf() {
+        ByteBufAdaptor bba = adaptor;
+        if (bba == null) {
+            ByteBufAllocatorAdaptor alloc = new ByteBufAllocatorAdaptor(
+                    BufferAllocator.heap(), BufferAllocator.direct());
+            return adaptor = new ByteBufAdaptor(alloc, this);
+        }
+        return bba;
+    }
+
+    @Override
+    public int refCnt() {
+        return isAccessible()? 1 + countBorrows() : 0;
+    }
+
+    @Override
+    public ReferenceCounted retain() {
+        return retain(1);
+    }
+
+    @Override
+    public ReferenceCounted retain(int increment) {
+        for (int i = 0; i < increment; i++) {
+            acquire();
+        }
+        return this;
+    }
+
+    @Override
+    public ReferenceCounted touch() {
+        return this;
+    }
+
+    @Override
+    public ReferenceCounted touch(Object hint) {
+        return this;
+    }
+
+    @Override
+    public boolean release() {
+        return release(1);
+    }
+
+    @Override
+    public boolean release(int decrement) {
+        for (int i = 0; i < decrement; i++) {
+            close();
+        }
+        return !isAccessible();
```

Comment:
I think this methods can lead up to some unexpected behaviour. It may throw while still closing the underlying resources if `decrement` is > as the number of acquires that were done before.